cnf-tests: initiate konflux dockerfile

shajmakh opened the pull request:
Add dockerfile for konflux builds production.

The Dockerfile can be further simplified but for now this is just the 1:1 mapping of m/s configuration to u/s.

shajmakh commented:
/cc @Tal-or @zeeke @yanirq @rbaturov

shajmakh commented:
note: rpm repos for installing required binaries that aren't packed in the ubi image will follow later as such configuration will mostly be added to the konflux ci project.

zeeke commented:
Do you have any pointers to docs explaining how this new Dockerfile is consumed?

I see there are mixed internal (brew.registry.redhat.io) and external (registry.redhat.io) repositories, is that correct?

Tal-or commented on cnf-tests/Dockerfile.konflux:
```diff
@@ -0,0 +1,74 @@
+FROM brew.registry.redhat.io/rh-osbs/openshift-ose-cli-rhel9:v4.19 AS oc
```
Is there a reason for drifting from `registry.ci.openshift.org` to `brew.registry.redhat.io`?

shajmakh replied:
`brew.registry.redhat.io` is the external avialable link for `registry-proxy`, this was intended because we are doing a mapping from midstream to u/s and in m/s `registry-proxy` was used

shajmakh commented:
> Do you have any pointers to docs explaining how this new Dockerfile is consumed?

shajmakh commented:
> Do you have any pointers to docs explaining how this new Dockerfile is consumed?
> 
> I see there are mixed internal (brew.registry.redhat.io) and external (registry.redhat.io) repositories, is that correct?

sure, resources: https://konflux.pages.redhat.com/docs/users/how-tos/creating.html#_onboarding_components_to_konflux

The mixed registries are expected, this is how it worked so far but instead of `brew.registry.redhat.io` we had `registry-proxy` and in fact `brew.registry.redhat.io` is a way to get access to internal content like `registry-proxy`: https://konflux.pages.redhat.com/docs/users/getting-started/components-applications.html#_notable_pull_secrets

shajmakh commented:
> 
> The mixed registries are expected, this is how it worked so far but instead of `brew.registry.redhat.io` we had `registry-proxy` and in fact `brew.registry.redhat.io` is a way to get access to internal content like `registry-proxy`: https://konflux.pages.redhat.com/docs/users/getting-started/components-applications.html#_notable_pull_secrets

@zeeke Please note that your exact point was raised (mixed registries internal and production) mainly regarding producing production builds and considering this was the way it is done so far we agreed that we start with the equivalent mapping as was in m/s and later when the basic build flow pass we start improving in case needed.

shajmakh commented:
konflux github app is needed
/hold

shajmakh commented:
/retest

shajmakh commented:
/retest

shajmakh commented:
This MR is reviewable now.

/unhold

swatisehgal commented on cnf-tests/.konflux/Dockerfile:
```diff
@@ -0,0 +1,74 @@
+COPY . .
+RUN make test-bin
+
+FROM registry.redhat.io/ubi9/ubi-minimal:9.5
```
The container image should be based on the same RHEL version as OCP. Let's use a `RHEL_VERSION`  to be able to override it in case needed. Please see https://github.com/openshift-kni/dpdk-base/pull/5 for reference.

OCP 4.19 GA will be based on RHEL 9.4 EUS so we should use RHEL version as 9.4 here.

shajmakh replied:
Thanks for highlighting and for the reference, I'll adopt this.

yanirq commented on .tekton/cnf-tests-pull-request.yaml:
@shajmakh  we are we adding the tekton files in advance ?  shouldn't they be generated by the konflux bot ?

shajmakh replied:
they were, I copied that from the PR https://github.com/openshift-kni/cnf-features-deploy/pull/2170 to be able to test the dockerfile

shajmakh replied:
we still can modify these files which is something I am looking into to allow storing images post-merge in different registries while adding CVP-like tests before the are published

zeeke commented:
@shajmakh I see job [Red Hat Konflux / cnf-tests-on-pull-request](https://github.com/openshift-kni/cnf-features-deploy/pull/2167/checks?check_run_id=36469361092) failed but I can't open the URL
https://console.redhat.com/application-pipeline/ns/telco-5g-tenant/pipelinerun/cnf-tests-on-pull-request-z52kf

as it says 
> Unable to load pipeline run cnf-tests-on-pull-request-z52kf

Is there anything on my side I should configure? Or is it a problem in the app configuration?

shajmakh commented:
@zeeke seems like an outage on the app end because I see the initialization of the build failed too. I see the failure is on cachi2 service (check https://github.com/openshift-kni/cnf-features-deploy/pull/2167/checks?check_run_id=36469361092) so let me try to retrigger
/retest
The latest updates modified the image tag that would be used upon storing the image in Quay where I added `4.19` prefix:
https://github.com/openshift-kni/cnf-features-deploy/pull/2167/files#diff-a2563ccd46641fcd2fbb49a6e3ed3b9eaec4f6d57a610ab774cce7ef6dda9c3cR26

https://github.com/openshift-kni/cnf-features-deploy/pull/2167/files#diff-5706bd43485225cba85ba2ac6583e6e9a85d97d2a3e03744d14afa42ddcb3dc4R25

shajmakh commented:
/hold
rethinking this I don't think it is the right thing to override the latest everytime a pr is merged. we need some release numbers added to 4.19 builds.

shajmakh commented:
> /hold rethinking this I don't think it is the right thing to override the latest everytime a pr is merged. we need some release numbers added to 4.19 builds.

A ticket was filed for the Konflux team to support cron jobs. I believe we can continue with the current configuration considering there is no better alternative. 
/unhold

shajmakh commented:
/retest

zeeke commented:
`cnf-tests/.konflux/redhat.repo` file is 8K+ lines. Do we need that? It sounds a little bit complicated to maintain.

zeeke commented on cnf-tests/.konflux/rpms.in.yaml:
```diff
@@ -0,0 +1,32 @@
+    # Alternative to setting command line options. Usually you will only want
+    # to include one of these options, with the exception of `flatpak` that
+    # can be combined with `image` and `containerfile`
+    containerfile: Containerfile
```
There is no `Containerfile` in this folder. Do we need to change this to `Dockerfile`?
Also, I'm not getting where this `rpms.in.yaml` file is referenced from. Can we get rid of this?

shajmakh replied:
The containerfile was local indeed and we needed rpms.in.yaml to generate the rpms.lock file, however konflux instructions is to upload these to the source repo. I'll double-verify anyway. Thanks for your comment.

shajmakh replied:
testing in #2209

shajmakh replied:
we could use any docker/container file that has the final image as ubi, I replaced this to use the cnf-tests Dockerfile. The process of updating the rpms.lockfile.yaml with more deps if needed will be ironed and automatic updates to lockfiles as per konflux team as we advance so it would be easier to do updates to the deps.

yanirq commented:
> `cnf-tests/.konflux/redhat.repo` file is 8K+ lines. Do we need that? It sounds a little bit complicated to maintain.

ubi.repo might be sufficient here. we can try and remove this.

shajmakh commented:
looks like a legit failure on the metallb tests:
```
{Timed out after 180.001s.
Unexpected error:
    <*errors.errorString | 0xc00028e070>: 
    daemonset speaker pod frr-k8s-2jstt is not running, expected status Running got Pending
    {
        s: "daemonset speaker pod frr-k8s-2jstt is not running, expected status Running got Pending",
    }
occurred failed [FAILED] Timed out after 180.001s.
Unexpected error:
    <*errors.errorString | 0xc00028e070>: 
    daemonset speaker pod frr-k8s-2jstt is not running, expected status Running got Pending
    {
        s: "daemonset speaker pod frr-k8s-2jstt is not running, expected status Running got Pending",
    }
occurred
In [It] at: /go/src/github.com/openshift-kni/cnf-features-deploy/cnf-tests/submodules/metallb-operator/test/e2e/functional/tests/e2e.go:280 @ 02/03/25 12:12:47.753
}
```

@zeeke Athough this PR doesn't touch the tests/product code and the lane is not required to merge the PR, do you perhaps know if this is something that the team is looking at from the network end?

shajmakh commented:
redhat.repo seems not needed for the rpm prefetching to work. I'm testing removing unnecessary files in #2209 and will confirm this here

rbaturov commented on cnf-tests/.konflux/Dockerfile:
```diff
@@ -0,0 +1,74 @@
+COPY --from=builder-sctptester /sctptest /usr/bin/sctptest
+COPY --from=builder-hugepages-allocator /hugepages-allocator /usr/bin/hugepages-allocator
+
+RUN sed -i 's/quay.io\/openshift-kni\//registry.redhat.io\/openshift4\//g' /usr/local/etc/cnf/images.json
+RUN sed -i 's/cnf-tests:4.19/cnf-tests-rhel9:v4.19/g' /usr/local/etc/cnf/images.json
+RUN sed -i 's/dpdk:4.19/dpdk-base-rhel8:v4.19/g' /usr/local/etc/cnf/images.json
+
+ENV IMAGE_REGISTRY=registry.redhat.io/openshift4/
+ENV CNF_TESTS_IMAGE=cnf-tests-rhel9:v4.19
+ENV DPDK_TESTS_IMAGE=dpdk-base-rhel8:v4.19
```
Let's update the hardcoded 4.19 values to use an env var like OCP_VERSION.
This is to maintain consistency with [this](https://github.com/openshift-kni/cnf-features-deploy/pull/2187/files#diff-1f7ab57f320c60bb215a880e6d077d84dbb9ac6bab1623584fe364e003a45285R87-R90). Thogh this could be addressed in a later PR, what works best for you.

shajmakh commented:
/retest

shajmakh commented:
Following the discussions here and offline, we are strongly recommended by the product security team to use hermetic build flow (network isolated):

>  meets security build best practices, protects against potential supply chain threats, gives us assurance that we know exactly what is included in the build

please see:https://konflux.pages.redhat.com/docs/users/faq/hermetic.html
following the resource, services can still use an exception to use non-hermetic (network connected) build flow but we need to ask for an exception, and for cnf-tests we need a fair reason to get an exception. We will sooner or later need to move to use hermetic builds way and starting from now sounds a better option to avoid disruption later in the process. 

 which means that dependencies are prefetched by a konflux build task which looks for the RPM lockfile to detect the right RPMs to download for the offline build. This is where rpms.lockfile.yaml is needed. However, because the lockfile expects even transitive dependencies resolved and listed it can be potentially massive. That's where the rpms.in.yaml comes in, that's a file where one should list only the direct dependencies and let other tool (in this case this one: https://github.com/konflux-ci/rpm-lockfile-prototype) generate the resolved lockfile for us. That said, this generator is unable to resolve all the RPM transitive dependencies correctly if one don't provide it with a source of truth, i.e. a .repo file.
 
That said, in order for konflux build flow to function as expected the network-isolated way, we need to to have lockfile, .repo file, and the rpms.in.yaml.

@yanirq @zeeke please let me know if you have additional concerns.

zeeke commented:
@shajmakh thanks for having explained the whole picture. Now, it looks way clearer to me.

I'm ok with having the `ubi.repo` as its size looks maintainable (vs 8k+ lines of the previous version).

/approve

LGTM (leaving the compute team time for reviewing this)

shajmakh commented:
cachi2 failure, try again
/retest

shajmakh commented:
known issue in ICM step to pull the image, a thread was filed. Try rerun again.
/retest

yanirq commented:
@shajmakh @rauhersu @rbaturov @fontivan   we need to find a configuration that will not apply all konflux CI lanes under a certin component  (here we see atm ztp tests running for example) 
Also , do we need **both** staging and production ecp lanes running here ?

shajmakh commented:
@zeeke @yanirq I have updated this patch further, please check the below details when you get the chance and let me know if you have any concerns. Thanks. 
Summary of the last push:
- dropped `ubi.repo` 
- readded simplified `redhat.repo`
- updated input file to point to redhat repo 

The reason why I dropped the ubi.repo is because it has 2 RPM repos enabled that are by default available in the ubi-minimal base, so there is actually no need for that. One woul'd probably ask "So what RPM repos did Konflux flow actually used?" I tell you `redhat.repo`! but it didn't need to be attached here because rpms.in.yaml was pointing to `ubi.repo` one so it was passing. 
The next question would likely be "then how did it know the source of truth of where to pull the rpms?" -> we have generated the lockfile correctly with the sourcerpm and repoid fields set as needed. The next question will likely be "okay, so that all was generated correctly locally, but we know that konflux regenerates the lockfile, how is it enabled to install the tools without having the right repos enabled in the `.repo` file? "
The answer to that hides within the fact that in konflux instance where these integration tests are configured and running, we have a place to store the secrets and activation keys safely, The (only one) activation key that is used there has rhel-baseos and rhel-appstream RPM repos enabled by default so when konflux runs all the flow again of regenerating the lockfile it subscribes with that key and can successfully install the tools.

yanirq commented on cnf-tests/.konflux/redhat.repo:
```diff
@@ -0,0 +1,27 @@
+[rhel-9-for-$basearch-appstream-beta-rpms]
```
Beta is the most stable version to be used here ?

shajmakh replied:
I noticed that as well, but this is the core repos added by default to the activation key used for development and testing and surprisingly I can't see the GA ones that I can add in the activation key web UI. I'll check if changing the type can help here.

shajmakh replied:
checked with all kinds of usages and all are having beta repos, in fact all the rhel bases are beta:
```
[root@023221282e9a yum.repos.d]# cat redhat.repo | grep rhel-9-for
[rhel-9-for-x86_64-sap-netweaver-beta-rhui-debug-rpms]
[rhel-9-for-x86_64-highavailability-beta-source-rpms]
[rhel-9-for-x86_64-sap-netweaver-beta-rpms]
[rhel-9-for-x86_64-resilientstorage-beta-source-rpms]
[rhel-9-for-x86_64-supplementary-beta-source-rpms]
[rhel-9-for-x86_64-sap-solutions-beta-rpms]
[rhel-9-for-x86_64-appstream-beta-rhui-source-rpms]
[rhel-9-for-x86_64-rt-beta-debug-rpms]
[rhel-9-for-x86_64-baseos-beta-debug-rpms]
[rhel-9-for-x86_64-appstream-beta-source-rpms]
[rhel-9-for-x86_64-rt-beta-rpms]
[rhel-9-for-x86_64-sap-solutions-beta-source-rpms]
[rhel-9-for-x86_64-appstream-beta-debug-rpms]
[rhel-9-for-x86_64-baseos-beta-source-rpms]
[rhel-9-for-x86_64-appstream-beta-rhui-rpms]
[rhel-9-for-x86_64-nfv-beta-source-rpms]
[rhel-9-for-x86_64-rt-beta-source-rpms]
[rhel-9-for-x86_64-baseos-beta-rpms]
[rhel-9-for-x86_64-supplementary-beta-rpms]
[rhel-9-for-x86_64-supplementary-beta-debug-rpms]
[rhel-9-for-x86_64-sap-netweaver-beta-rhui-source-rpms]
[rhel-9-for-x86_64-appstream-beta-rpms]
[rhel-9-for-x86_64-highavailability-beta-rpms]
[rhel-9-for-x86_64-nfv-beta-rpms]
[rhel-9-for-x86_64-sap-solutions-beta-rhui-source-rpms]
[rhel-9-for-x86_64-sap-netweaver-beta-rhui-rpms]
[rhel-9-for-x86_64-baseos-beta-rhui-debug-rpms]
[rhel-9-for-x86_64-appstream-beta-rhui-debug-rpms]
[rhel-9-for-x86_64-sap-netweaver-beta-source-rpms]
[rhel-9-for-x86_64-sap-solutions-beta-rhui-rpms]
[rhel-9-for-x86_64-sap-solutions-beta-rhui-debug-rpms]
[rhel-9-for-x86_64-nfv-beta-debug-rpms]
[rhel-9-for-x86_64-baseos-beta-rhui-rpms]
[rhel-9-for-x86_64-sap-netweaver-beta-debug-rpms]
[rhel-9-for-x86_64-baseos-beta-rhui-source-rpms]
[rhel-9-for-x86_64-resilientstorage-beta-rpms]
[rhel-9-for-x86_64-highavailability-beta-debug-rpms]
[rhel-9-for-x86_64-sap-solutions-beta-debug-rpms]
[rhel-9-for-x86_64-resilientstorage-beta-debug-rpms]
[root@023221282e9a yum.repos.d]# cat redhat.repo | grep rhel-9-for| grep -v beta
[root@023221282e9a yum.repos.d]# 
```

shajmakh replied:
I've opened an IT ticket to calrify this. Will update once I have any news.

shajmakh replied:
It turns out that using a production base image affects which kind of repos will be displayed under the .repo file (beta/ga). I regenerated the lockfile with GA repos using a production ubi image.

yanirq commented on cnf-tests/.konflux/rpms.lock.yaml:
```diff
@@ -0,0 +1,57 @@
+---
```
Maybe its worth adding a readme file / short script in the meanwhile to show how to update this file ?

shajmakh replied:
Generating it can be done following the steps in the docs, adding a link to that is handy and a good idea. I see the point that updating RPMs deps is a concern but this file shouldn't never be updated manually or else runs will very likely fail if updates were missed for the input file and the repo. ADd to that that updating the repo file will need adjustments to the activation key reference in konflux instance and the one used to locally generate the lockfile.

shajmakh replied:
I have added a readme  file in the last commit, please consider this will be adjusted once we have settled on an activation key.

zeeke commented:
@shajmakh thanks again for having explained the underline logics. 

I'm fine with the `redhat.repo` file. As long as it is small and maintainable, it's good.

+1 on yanir proposal to add comment/script to update `.lock` file. Then LGTM

cgoncalves commented on cnf-tests/.konflux/redhat.repo:
```diff
@@ -0,0 +1,27 @@
+[rhel-9-for-$basearch-appstream-beta-rpms]
```
For RHEL EUS releases like 9.4 EUS, the EUS RPM repositories should also be included no?
Otherwise the container may not pick up the latest CVE fixes.

shajmakh replied:
good question, but do we want to build the image with EUS base? how do we know that?

cgoncalves replied:
There is no EUS base image. There is only latest 9.4 which is non-EUS to the best of my knowledge. That is why I think we have to find a way to include the EUS repositories (when applicable) and maybe even run a 'dnf update' so that we get from 9.4 to 9.4 EUS plus installing whichever RPMs we need from EUS repositories.

shajmakh replied:
I have added the EUS repos but for now it fails to install the packages for some reason that is why we don't see that in the lockfile. I believe this step is anyway advanced and should not block this PR considering we are speaking about extended update support for rhel. Also konflux bot will be periodically updating the deps automatically including the bases and soon lockfiles. We can file a card on the konclux side to support aligning our operators with the OCP deps using some param or something similar to indicate that.

shajmakh replied:
EUS repo support:
https://github.com/openshift-kni/cnf-features-deploy/pull/2285

cgoncalves commented on .tekton/cnf-tests-pull-request.yaml:
```diff
@@ -0,0 +1,620 @@
+    - name: dockerfile
+      value: cnf-tests/.konflux/Dockerfile
+    - name: prefetch-input
+      value: '{"type": "rpm", "path": "cnf-tests/.konflux"}'
```
I think an explicit hermetic setting must be set to ensure network isolation:

```suggestion
      value: '{"type": "rpm", "path": "cnf-tests/.konflux"}'
    - name: hermetic
      value: "true"
```

See https://konflux.pages.redhat.com/docs/users/how-tos/configuring/hermetic-builds.html

yanirq replied:
will conforma issue an error if  hermetic is not explicitly set here ?

cgoncalves replied:
I'm not sure. Hermetic builds are strongly recommended but not mandatory from what I read.

fontivan replied:
I believe conforma will fail if hermetic is not enabled

shajmakh replied:
Thanks for highlighting Carlos. I will fix it. I wonder how come this was accepted so far by the CI without this param, perhaps they newly added it.

cgoncalves commented on cnf-tests/.konflux/redhat.repo:
```diff
@@ -0,0 +1,27 @@
+[rhel-9-for-$basearch-appstream-rpms]
+name = Red Hat Enterprise Linux 9 for $basearch - AppStream (RPMs)
+baseurl = https://cdn.redhat.com/content/dist/rhel9/$releasever/$basearch/appstream/os
```
`$releasever` will resolve to `9` which means the latest UBI 9 version (UBI 9.5). You will have a mix of 9.4 RPMs from the base image and 9.5 RPMs from the RPMs you want to additionally install. You can see those 9.5 RPMs in the `rpms.lock.yaml` file.

shajmakh replied:
good comment and I agree. I'm addressing this in #2285  and waiting for a recommended way to pass the full x.y stream version while manually generating this and during the prefetch step by the integration test.

shajmakh replied:
https://github.com/openshift-kni/cnf-features-deploy/pull/2285/commits/005a9d5ac75a777fe5a786931f604233fa868c73#diff-820b98e7bc57b3ec683c00a0fb7e167c32796d66242b25c2ace5d30427bbb222R17

fontivan commented:
/lgtm

shajmakh commented:
production and staging conforma (FKA EC or Enterprise Contract) shouldn't be blocking PRs if `cnf-tests-4-19-on-pull-request` passed. Conforma policy adjustments will be made on the konflux release data to adjust the rules.
 
/override "Red Hat Konflux / cnf-tests-ec-production-4-19 / cnf-tests-4-19"
/override "Red Hat Konflux / cnf-tests-ec-staging-4-19 / cnf-tests-4-19"

zeeke commented:
/override "Red Hat Konflux / cnf-tests-ec-production-4-19 / cnf-tests-4-19"
/override "Red Hat Konflux / cnf-tests-ec-staging-4-19 / cnf-tests-4-19"
/test e2e-aws-ci-tests

shajmakh commented:
/test e2e-aws-ci-tests

shajmakh commented:
`Red Hat KonfluxQueued — Pending approval` is a known issue https://issues.redhat.com/browse/SRVKP-7242

shajmakh commented:
> `Red Hat KonfluxQueued — Pending approval` is a known issue https://issues.redhat.com/browse/SRVKP-7242

depends on https://github.com/openshift-pipelines/pipelines-as-code/pull/1982

shajmakh commented:
/retest

ralphbean commented on cnf-tests/.konflux/README.md:
```diff
@@ -0,0 +1,40 @@
+
+To delete no longer needed dependencies you need to:
+1. remove the binary name from `rpms.in.yaml` and from `Dockerfile`.
+2. remove the matching section for this binary from `rpms.lock.yaml`.
```
If preferred, you could re-run `rpm-lockfile-prototype` to regenerate it from `rpms.in.yaml`, instead of manually editing it.

shajmakh replied:
you're right, I mentioned this option in https://github.com/openshift-kni/cnf-features-deploy/pull/2167/files/a46c342c9b845c1b66902ed9046e49e8c193c036#diff-c08a78341751a7ef28dd58f831921ab6668ce1c20a4327fd2b1417baa99bef13R13

cgoncalves commented:
/lgtm

yanirq commented on cnf-tests/.konflux/README.md:
```diff
@@ -0,0 +1,40 @@
+**Update RPMs dependencies**
```
@ralphbean  , Perhaps this README can be a good addition to the konflux docs as well ?

shajmakh commented:
conforma are not blocking at this stage as per:
https://github.com/openshift/release/pull/62780
but I don't have permissions to override.
@zeeke can you help override them again please?

zeeke commented:
> conforma are not blocking at this stage as per: [openshift/release#62780](https://github.com/openshift/release/pull/62780) but I don't have permissions to override. @zeeke can you help override them again please?

sounds good

/override Red Hat Konflux / cnf-tests-ec-production-4-19 / cnf-tests-4-19

zeeke commented:
/override "Red Hat Konflux / cnf-tests-ec-production-4-19 / cnf-tests-4-19"

zeeke commented:
/override "Red Hat Konflux / cnf-tests-ec-staging-4-19 / cnf-tests-4-19"

fontivan commented:
/override "Red Hat Konflux / ztp-site-generate-conforma-production-4-19 / ztp-site-generate-4-19"

zeeke commented:
> /override "Red Hat Konflux / ztp-site-generate-conforma-production-4-19 / ztp-site-generate-4-19"


/override "Red Hat Konflux / ztp-site-generate-conforma-production-4-19 / ztp-site-generate-4-19"
/override "Red Hat Konflux / ztp-site-generate-conforma-staging-4-19 / ztp-site-generate-4-19"
